### ISSUE TITLE
Provide an API to create extensions for the devtools

### ISSUE THREAD

**Issue by simc** [flutter/devtools#1632]
It would be very helpful if there was a way to create custom extensions for the devtool. For example, I'd like to write an extension to inspect the contents of the [Hive DB](/hivedb/hive). But I think there are many interesting extensions, the community could create.

The Stetho tool from Facebook allows [something similar](https://github.com/facebook/stetho#custom-dumpapp-plugins) for the Chrome DevTools.

And the flipper tool also has a very powerful way to [create extensions](https://fbflipper.com/docs/tutorial/intro.html). It even allows creating [custom UI](https://fbflipper.com/docs/tutorial/js-custom.html).

**Comment by britannio** [comment]
Apologies for the useless comment but this is my no. 1 request.
Few use cases
- Monitoring network activity for a network package e.g. dio or graphql
- Tracking state changes in Redux
- Creating custom dashboards with key metrics for your app
- Inspecting databases as @leisim suggested
- ...

**Comment by cedvdb** [comment]
The use cases are legion tbh, any kind of inspection of state is useful there. Be it db state, application state, ui state, cache state, whatever state. Those kind of things are impractical to print inside the console because:

 - It's usually pretty big and would be unreadable
 - It usually involve a tree structure 

I would also like the possibility to have extension, but if that's not planned if there was a message bus mechanism where one could start another app that connects to that bus and display the required information with a desktop app it would be nice.

Let me rephrase this, if extensions is not planned, it would be nice to have the ability to:
  - start a flutter app that emits messages.
  - start another app that displays those messages.

Maybe this is doable today but my guess is that the process will be too involved (development wise) to make it practical.

**Comment by leisim** [comment]
It would even be helpful to allow third parties to access an active debugging session without requiring the user to find the websocket link in the logs.

Use cases:
- [Isar Inspector](https://github.com/isar/isar#isar-inspector)
- Custom integration test VSCode extension

**Comment by leisim** [comment]
@kenzieschmoll Is there something we could help with to make this happen? Or are there alternatives? Riverpod and Provider are integrated into devtools. Can anyone submit an extension?

**Comment by jacob314** [comment]
Here's a link to our in progress design on supporting Flutter DevTools Plugins.  Comments appreciated! https://docs.google.com/document/d/1BWX8YQ962Vsx-EUDuDHRG7RX94fJlZSDRu299YRwasE/edit

**Comment by leisim** [comment]
@jacob314 Wow that is incredible 😍 can't wait to build some plugins

**Comment by rrousselGit** [comment]
:wave:
Out of curiosity, when could this be made available and is there anything we can do to help? :pray:

I'm asking because I'm honestly considering making my own plugin mechanism through a custom VScode extension which would allow folks to contribute tabs.   
I was considering a VScode extension because it enables the plugin mechanism to access the full VScode API.  
So plugins could provide custom commands or implement their own "open a file at a given offset".

But if we're going to have plugins in a reasonable timeframe, this project may not be worth it anymore (since an official solution would have a bigger reach)

**Comment by jacob314** [comment]
@kenzieschmoll is in the process of kicking this effort back off. VSCode API integration is interesting and we have use cases like editing Flutter widget properties where that would be helpful as well. It would be great to have you collaborate on the implementation of the official solution for this.

**Comment by rrousselGit** [comment]
It's a bit unclear to me on what I could do to help. Do you have anything more specific?

Honestly, I wouldn't mind contributing to various aspects of the plugin mechanism. VScode API or not.

**Comment by kenzieschmoll** [comment]
Hi @rrousselGit. I'm going to prioritize working on this design. I will keep you posted when I have some more specific details to share. Thanks!

**Comment by kenzieschmoll** [comment]
I've been working on a prototype to provide a proof of concept for the above designs.

Screencast link: https://screencast.googleplex.com/cast/NTk2MzgyOTQ4MjYxODg4MHxiODFhYjFiOC0yNA
(or download video): [DevTools Plugins Prototype.webm](https://github.com/flutter/devtools/assets/43759233/80c8394e-2db8-4d2a-bc4d-5f5ac651b55c)

This work is subject to change and needs a lot of cleanup, but early & rough prototype available with the combination of the following changes:

- DevTools: https://github.com/kenzieschmoll/devtools/tree/tooling-extensions
- DevTools Server (package:dds in the dart SDK): https://dart-review.googlesource.com/c/sdk/+/305043. 

For testing, I am running `app_that_uses_foo` from here, which imports the `foo_package` from the same repo: https://github.com/kenzieschmoll/test_apps/tree/main/tooling_plugins, and I am connecting that app to devtools, built and served with `dart devtools/tool/build_e2e.dart`.

**Note**: for this to work properly, the `app_that_uses_foo/.dart_tool/package_config.json` entry for "foo_package" needs to use an absolute path, not a relative one:
```
    {
      "name": "foo_package",
      "rootUri": "../../foo_package", // Change to an absolute path
      "packageUri": "lib/",
      "languageVersion": "3.1"
    },
```

**Comment by kenzieschmoll** [comment]
DevTools app
- [ ] Conditionally show screens when extensions are available - done in https://github.com/flutter/devtools/pull/6048
- [ ] Implement extension screen framework in DevTools with embedded iFrame - done in https://github.com/flutter/devtools/pull/6054, https://github.com/flutter/devtools/pull/6060, [more to come]
- [ ] Implement extension detection through DevTools server
- [ ] Implement extension activation through DevTools server

DevTools server
- [ ] develop a package to detect available extensions based on the `.dart_tool/package_config.json` file for an app @jonasfj 
- [ ] use the extension detection package in devtools_server to find available extensions
- [ ] add logic to read / write extension activation states from a `devtools_options.yaml` file

package:devtools_extensions
- [ ] create package - done in https://github.com/flutter/devtools/pull/6055
- [ ] create the shared api that both DevTools and DevTools extensions will use
- [ ] create a template that extension authors will use to create devtools extensions

extensions land well
- [ ] Write the first DevTools extension and publish (package:provider?)
- [ ] Write documentation for extension authors and publicize (tweet, medium article, etc.)
- [ ] Gather early adopters to write extensions and provide feedback

**Comment by jonasfj** [comment]
Initial draft for a `package:extension_discovery` in https://github.com/dart-lang/tools/pull/129

**Comment by rrousselGit** [comment]
Cool! Tell me whenever there's something I can play with :)

**Comment by kenzieschmoll** [comment]
This work has landed! Anyone wanting to write a DevTools extension should follow the instructions at https://pub.dev/packages/devtools_extensions. Extension support in DevTools just landed in the Dart SDK and should be available on Flutter master by the end of the week. This will be included in the next Flutter beta and stable releases. In the meantime, you can still develop your extension using the simulated DevTools environment described in the [instructions](https://pub.dev/packages/devtools_extensions#debug-the-extension-web-app). 

`package:provider` has already published a DevTools extension for version [6.1.0-dev.1](https://pub.dev/packages/provider/versions/6.1.0-dev.1). We are working on setting up a Discord channel specific to DevTools extension authors, but in the meantime you can reach out on Github or on the #hackers-devtools channel on the [Flutter discord](https://github.com/flutter/flutter/wiki/Chat).

We are eager to hear early feedback on this feature and your input on how we can improve it!
